### PR TITLE
[IPv6] Add CBMC test for xProcessReceivedTCPPacket_IPv4 & xProcessReceivedTCPPacket_IPv6

### PR DESCRIPTION
**Pull request by ActoryOu**
[IPv6] Add CBMC test for xProcessReceivedTCPPacket_IPv4 & xProcessReceivedTCPPacket_IPv6

Description
-----------

- Add CBMC test cases for the functionxProcessReceivedUDPPacket_IPv6.
- Add IPv4 in original xProcessReceivedUDPPacket test case.
  - These test cases (both v4 and v6) also verifies below functions.
    - xProcessReceivedTCPPacket
    - xProcessReceivedTCPPacket_IPV4
    - xProcessReceivedTCPPacket_IPV6
    - prvTCPSendReset
    - prvTCPSendSpecialPacketHelper
    - prvTCPSendSpecialPktHelper_IPV4
    - prvTCPSendSpecialPktHelper_IPV6
    - prvTCPSendRepeated
    - prvTCPSendChallengeAck

Test Steps
-----------

```
cd test/cbmc/proof/
python3 prepare.py
cd test_case
make
```

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/Makefile.json [hunk 1]
```diff
@@ -9,10 +9,12 @@
   "OBJS":
   [
     "$(ENTRY)_harness.goto",
+    "$(FREERTOS_PLUS_TCP)/test/cbmc/stubs/cbmc.goto",
```
Comment (moninom1) [inline test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/Makefile.json:12]: Why is this needed?

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket_IPv6/Makefile.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    "--unwindset prvTCPSendRepeated.0:13",
+    "--nondet-static"
+  ],
+  "OBJS":
+  [
+    "$(ENTRY)_harness.goto",
+    "$(FREERTOS_PLUS_TCP)/test/cbmc/stubs/cbmc.goto",
+    "$(FREERTOS_PLUS_TCP)/source/FreeRTOS_IP.goto",
+    "$(FREERTOS_PLUS_TCP)/source/FreeRTOS_IP_Utils.goto",
+    "$(FREERTOS_PLUS_TCP)/source/FreeRTOS_IP_Timers.goto",
```
Comment (moninom1) [inline test/cbmc/proofs/parsing/ProcessReceivedTCPPacket_IPv6/Makefile.json:15]: I doubt we need all these includes, can we also mention the functions that are being validated as part of each test in the commit message? It will give idea of which includes are really needed.

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/Makefile.json (hunk 1 above)
Comment (ActoryOu) [reply test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/Makefile.json:12]: To get the source of [safeMalloc](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/dev/IPv6_integration/test/cbmc/stubs/cbmc.c#L8).

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c [hunk 3]
```diff
@@ -52,6 +42,40 @@ TaskHandle_t xTaskGetCurrentTaskHandle( void )
...
+        if( xRetSocket->u.xTCP.pxPeerSocket == NULL )
+        {
+            xRetSocket->u.xTCP.bits.bReuseSocket = pdTRUE_UNSIGNED;
+        }
+        else
+        {
+            xRetSocket->u.xTCP.bits.bReuseSocket = pdFALSE_UNSIGNED;
+        }
+
+        if( xIsCallingFromIPTask() == pdFALSE )
```
[the remark this answers is not in the document]
Comment (ActoryOu) [reply test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c:69]: Great catch. Let me update it.

[the remark this answers is not in the document]
Comment (ActoryOu) [reply test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c]: Updated

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket_IPv6/Makefile.json (hunk 2 above)
Comment (ActoryOu) [reply test/cbmc/proofs/parsing/ProcessReceivedTCPPacket_IPv6/Makefile.json]: Remove redundant goto files. And update description of test case for sub-functions also validated.

**Review by tony-josi-aws — APPROVED** [review]

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c [hunk 4]
```diff
@@ -52,6 +42,40 @@ TaskHandle_t xTaskGetCurrentTaskHandle( void )
...
+        xRetSocket->u.xTCP.txStream = safeMalloc( sizeof( StreamBuffer_t ) );
+        xRetSocket->u.xTCP.pxPeerSocket = safeMalloc( sizeof( StreamBuffer_t ) );
+
+        /* This test case is for IPv4. */
+        __CPROVER_assume( xRetSocket->bits.bIsIPv6 == pdFALSE );
+
+        /* This bit depicts whether the socket was supposed to be reused or not. */
+        if( xRetSocket->u.xTCP.pxPeerSocket == NULL )
+        {
+            xRetSocket->u.xTCP.bits.bReuseSocket = pdTRUE_UNSIGNED;
```
Comment (moninom1) [inline test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c:62]: Why do we need to set it?

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c [hunk 5]
```diff
@@ -52,6 +42,40 @@ TaskHandle_t xTaskGetCurrentTaskHandle( void )
...
+        }
+        else
+        {
+            xRetSocket->u.xTCP.bits.bReuseSocket = pdFALSE_UNSIGNED;
+        }
+
+        if( nondet_bool() )
+        {
+            xRetSocket->u.xTCP.bits.bPassQueued = pdFALSE_UNSIGNED;
```
Comment (moninom1) [inline test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c:71]: Same why are we explicitly setting the value? CBMC should be able to generate all possible cases and cover it

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c [hunk 6]
```diff
@@ -52,6 +42,40 @@ TaskHandle_t xTaskGetCurrentTaskHandle( void )
     return pxCurrentTCB;
 }
 
+/* Abstraction of prvHandleListen_IPV4 */
+FreeRTOS_Socket_t * prvHandleListen_IPV4( FreeRTOS_Socket_t * pxSocket,
+                                          NetworkBufferDescriptor_t * pxNetworkBuffer )
+{
+    FreeRTOS_Socket_t * xRetSocket = safeMalloc( sizeof( FreeRTOS_Socket_t ) );
```
Comment (moninom1) [inline test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c:49]: Please use ensure_FreeRTOS_Socket_t_is_allocated for consistency

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c (hunk 4 above)
Comment (ActoryOu) [reply test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c]: It's useless. Removed.

File: test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c (hunk 6 above)
Comment (ActoryOu) [reply test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c:41]: It's possible to get NULL return. And it's impossible to get NULL from ensure_FreeRTOS_Socket_t_is_allocated.

Comment (moninom1) [reply test/cbmc/proofs/parsing/ProcessReceivedTCPPacket/ProcessReceivedTCPPacket_harness.c:41]: ensure_FreeRTOS_Socket_t_is_allocated can return NULL as well.

**Review by shubnil — APPROVED** [review]

**Review by shubnil — APPROVED** [review]
